src/audio/volume/volume.c

```diff
@@ -51,9 +50,138 @@ static const struct comp_driver comp_volume;
 DECLARE_SOF_UUID("volume", volume_uuid, 0xb77e677e, 0x5ff4, 0x4188,
 		 0xaf, 0x14, 0xfb, 0xa8, 0xbd, 0xbf, 0x86, 0x82);
 
-/* 34dc0385-fc2f-4f7f-82d2-6cee444533e0 */
-DECLARE_SOF_UUID("volume-task", volume_task_uuid, 0x34dc0385, 0xfc2f, 0x4f7f,
-		 0x82, 0xd2, 0x6c, 0xee, 0x44, 0x45, 0x33, 0xe0);
+#if CONFIG_FORMAT_S16LE
+/**
+ * \brief Used to find nearest zero crossing frame for 16 bit format.
+ * \param[in,out] source Source buffer.
+ * \param[in] frames Number of frames.
+ */
+static uint32_t vol_zc_get_s16(const struct audio_stream *source,
+			       uint32_t frames)
+{
+	uint32_t buff_frag = frames * source->channels - 1;
+	uint32_t curr_frames = frames;
+	int32_t prev_sum = 0;
+	uint32_t channel;
+	int16_t *src;
+	int32_t sum;
+	uint32_t i;
+
+	for (i = 0; i < frames; i++) {
+		sum = 0;
+
+		for (channel = 0; channel < source->channels; channel++) {
+			src = audio_stream_read_frag_s16(source, buff_frag);
+			sum += *src;
+			buff_frag--;
+		}
+
+		/* first sign change */
+		if (i > 0 && (sum ^ prev_sum) < 0)
```

Comment:
You could take i = 0 iteration to happen before this for loop to avoid this test.

Reply:
Or maybe better to make prev_sum persistent to one opportunity more for ZC.